final?)

yulia1Volkova opened the pull request.

yulia1Volkova committed:
Sprint 23 3

yulia1Volkova committed:
Sprint 23 step3 done

yulia1Volkova committed:
Sprint 23 step4 done

yulia1Volkova committed:
Sprint 23 step4 done теперь точно))

yulia1Volkova committed:
Sprint 23 step5 done

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/db/data/PlaylistRepositoryImpl.kt:
```diff
@@ -23,13 +29,107 @@ class PlaylistRepositoryImpl(
+        val plList = appDatabase.playlistDao().getPlaylists()
+        for (item in plList) {
+            if (item.tracks_id != "") {
+                val idTracks = playlistDbConvertor.mapToList(item.tracks_id) as MutableList
```
Все касты лучше делать безопасными, чтобы не было крешей:
```suggestion
                val idTracks = playlistDbConvertor.mapToList(item.tracks_id) as? MutableList
```

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/db/data/PlaylistRepositoryImpl.kt:
```diff
@@ -23,13 +29,107 @@ class PlaylistRepositoryImpl(
 
+    override suspend fun deletePL(playlist: Playlist) {
+        val tracksForDelete= playlistDbConvertor.mapToList(playlist.tracks_id)
+        GlobalScope.launch {
```
`GlobalScope` не рекомендуется использовать. У тебя функция помечена `suspend`, здесь не обязательно запускать корутину и скоуп не нужен

yulia1Volkova replied:
У меня почему-то без этого глобал скоуп только плейлист удаляется, а треки из него не удаляются((

ArturNurtdinov replied:
Возможно, что после вызова `appDatabase.playlistDao().deletePL(playlistDbConvertor.map(playlist))` условие `checkingPlaylists(item)` не отрабатывает корректно, потому что ты удалила из БД плейлист

ArturNurtdinov replied:
Попробуй убрать скоуп и сначала удалить треки, которые нужно удалить, а только после этого почистить плейлист из БД

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/media/ui/activity/CurrentPlaylistFragment.kt:
```diff
@@ -0,0 +1,313 @@
+    }
+
+    lateinit var pl: Playlist
+    lateinit var tracks: List<Track>
```
все данные необходимо хранить во вьюмодели

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/media/ui/activity/CurrentPlaylistFragment.kt:
```diff
@@ -0,0 +1,313 @@
+                .load(uri)
+                .placeholder(R.drawable.placeholder)
+                .centerCrop()
+                .transform(RoundedCorners(8))
```
В конструктор `RoundedCorners` нужно передавать кол-во пикселей. Это значит, что предварительно число, передаваемое в конструктор, нужно сконвертировать в пиксели, если мы хотим добиться одинаково скругления на девайсах с разной плотностью экрана

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/media/ui/activity/CurrentPlaylistFragment.kt:
```diff
@@ -0,0 +1,313 @@
+
+        val count = playlist.tracksCount
+        val trackPlural = this.resources.getQuantityString(plurals_2, count, count)
+        binding.trackCountTextView.text = "$trackPlural"
```
Здесь вроде можно просто
```suggestion
        binding.trackCountTextView.text = trackPlural
```

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/media/ui/activity/EditPLFragment.kt:
```diff
@@ -0,0 +1,96 @@
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        binding.createPlaylistButton.text = "Сохранить"
+        binding.newPLTextView.text = "Редактировать"
```
Все строчки необходимо выносить в ресурсы strings.xml

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/media/ui/activity/EditPLFragment.kt:
```diff
@@ -0,0 +1,96 @@
+    private fun getImage(playlist: Playlist): Uri? {
+        var uri: Uri?
+        if (playlist.uri != null) {
+            val filePath =
+                File(requireContext().getExternalFilesDir(Environment.DIRECTORY_PICTURES), "PL")
+            val file = File(filePath, playlist.uri)
+            uri = file!!.toUri()
```
Работу с файлами лучше не во фрагменте делать, а в репозитории, т.к. это к хранению данных относится, то есть к data слою

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/media/ui/activity/FavoriteFragment.kt:
```diff
@@ -111,4 +111,8 @@ class FavoriteFragment : Fragment(), TrackAdapter.TrackClickListener {
     }
+
+    override fun onLongClick(track: Track) {
+        TODO("Not yet implemented")
```
туду лучше убрать, иначе будет креш

ArturNurtdinov reviewed (commented):
1. Список треков на экране плейлиста не соответствует макетам:
![image](https://github.com/yulia1Volkova/PlaylistMaker/assets/49075521/f9d5d214-fb76-49b1-81c6-098894468dca)
Макет:
![image](https://github.com/yulia1Volkova/PlaylistMaker/assets/49075521/21449651-ff13-4637-a5a2-5bf1cfa581f3)
В данном случае слишком большой отступ от двух кнопок и не углы не скруглены
2. В тёмной теме нужно поправить цвет текста в диалогах:
![image](https://github.com/yulia1Volkova/PlaylistMaker/assets/49075521/204bf6b1-fa46-4a3c-8560-f4b5eb753c5e)
Достаточно будет только цвет текста поправить, сам диалог можно оставить как есть. Но поправить нужно во всех диалогах (при удалении плейлиста, трека из плейлиста и на экране создания плейлиста)
3. Поймал баг, что в плейлисте, где есть 1 трек, список не отображается:
![image](https://github.com/yulia1Volkova/PlaylistMaker/assets/49075521/15f3988e-8285-4423-9ce2-4d5d1c1bcae5)

yulia1Volkova committed:
Sprint 23 step5 done

Отредактировала по критическм замечаниеям вроде все что удалось воспроизвести)

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/media/ui/activity/EditPLFragment.kt:
```diff
@@ -30,6 +32,8 @@ class EditPLFragment : CreaatePLFragment() {
     }
 
     lateinit var currentPL: Playlist
+    var newUri=false
+    private var uriDb: String? = null
```
Данные лучше хранить во вьюмодели, иначе при пересоздании фрагмента они потеряются

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/media/ui/activity/CreaatePLFragment.kt:
```diff
@@ -28,14 +28,14 @@ import org.koin.androidx.viewmodel.ext.android.viewModel
 import java.io.FileOutputStream
 
-class CreaatePLFragment : Fragment() {
+open class CreaatePLFragment : Fragment() {
```
В этом классе используется `requireActivity().onBackPressedDispatcher.addCallback`, этот колбэк срабатывает даже когда экран уже уничтожен. Тебе нужно вызывать `remove` на добавленном колбэке в методе `onDestroyView`, чтобы пофиксить креш и утечку памяти

yulia1Volkova committed:
Sprint 23 step5 done

Отредактировано удаление плейлиста

yulia1Volkova committed:
Sprint 23 step5 done

Я сначала сделала много всякой ерунды, а потом увидела комментарий в PR как пофиксить)) надеюсь получилось, и теперь все ок!

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/media/ui/activity/CurrentPlaylistFragment.kt:
```diff
@@ -178,6 +179,12 @@ class CurrentPlaylistFragment : Fragment(), TrackAdapter.TrackClickListener,
+            override fun handleOnBackPressed() {
+                 findNavController().navigateUp()
+            }
+        })
```
Этот колбэк тоже нужно удалять в `onDestroyView`, чтобы не было такой же проблемы с крешами как в `CreatePlFragment`

ArturNurtdinov commented on app/src/main/java/com/practicum/playlistmaker/media/ui/activity/EditPLFragment.kt:
```diff
@@ -93,15 +94,24 @@ class EditPLFragment : CreaatePLFragment() {
+                findNavController().navigateUp()
+
+            }
+        })
```
И здесь тоже нужно добавленный колбэк очищать, исправь пожалуйста, т.к. креши сейчас происходят

yulia1Volkova committed:
Sprint 23 step5 done

Я сначала сделала много всякой ерунды, а потом увидела комментарий в PR как пофиксить)) надеюсь получилось, и теперь все ок!
